tfhe/src/core_crypto/gpu/entities/lwe_ciphertext.rs

```diff
@@ -0,0 +1,147 @@
+
+        LweCiphertextList::from_container(
+            container,
+            self.lwe_dimension().to_lwe_size(),
+            self.ciphertext_modulus(),
+        )
+    }
+
+    pub fn from_lwe_ciphertext<C: Container<Element = T>>(
+        h_ct: &LweCiphertext<C>,
+        streams: &CudaStreams,
+    ) -> Self {
+        let lwe_dimension = h_ct.lwe_size().to_lwe_dimension();
+        let lwe_ciphertext_count = LweCiphertextCount(1);
+        let ciphertext_modulus = h_ct.ciphertext_modulus();
+
+        // Copy to the GPU
+        let mut d_vec = CudaVec::new(lwe_dimension.to_lwe_size().0, streams, 0);
+        unsafe {
+            d_vec.copy_from_cpu_async(h_ct.as_ref(), streams, 0);
+        }
+        streams.synchronize();
+
+        let cuda_lwe_list = CudaLweList {
+            d_vec,
+            lwe_ciphertext_count,
+            lwe_dimension,
+            ciphertext_modulus,
+        };
+        Self(cuda_lwe_list)
+    }
+
+    pub fn into_lwe_ciphertext(&self, streams: &CudaStreams) -> LweCiphertext<Vec<T>> {
+        let lwe_ct_size = self.0.lwe_dimension.to_lwe_size().0;
+        let mut container: Vec<T> = vec![T::ZERO; lwe_ct_size];
+
+        unsafe {
+            self.0
+                .d_vec
+                .copy_to_cpu_async(container.as_mut_slice(), streams, 0);
+        }
+        streams.synchronize();
+
+        LweCiphertext::from_container(container, self.ciphertext_modulus())
+    }
+
+    /// ```rust
+    /// use tfhe::core_crypto::gpu::lwe_ciphertext_list::CudaLweCiphertextList;
+    /// use tfhe::core_crypto::gpu::CudaStreams;
+    /// use tfhe::core_crypto::prelude::{
+    ///     CiphertextModulus, LweCiphertextCount, LweCiphertextList, LweSize,
+    /// };
+    ///
+    /// let mut streams = CudaStreams::new_single_gpu(0);
+    ///
+    /// let lwe_size = LweSize(743);
+    /// let ciphertext_modulus = CiphertextModulus::new_native();
+    /// let lwe_ciphertext_count = LweCiphertextCount(2);
+    ///
+    /// // Create a new LweCiphertextList
+    /// let lwe_list = LweCiphertextList::new(0u64, lwe_size, lwe_ciphertext_count, ciphertext_modulus);
+    ///
+    /// // Copy to GPU
+    /// let d_lwe_list = CudaLweCiphertextList::from_lwe_ciphertext_list(&lwe_list, &mut streams);
+    /// let d_lwe_list_copied = d_lwe_list.duplicate(&mut streams);
+    ///
+    /// let lwe_list_copied = d_lwe_list_copied.to_lwe_ciphertext_list(&mut streams);
+    ///
+    /// assert_eq!(lwe_list, lwe_list_copied);
+    /// ```
+    pub fn duplicate(&self, streams: &CudaStreams) -> Self {
+        let lwe_dimension = self.lwe_dimension();
+        let ciphertext_modulus = self.ciphertext_modulus();
+
+        // Copy to the GPU
+        let d_vec = unsafe { self.0.d_vec.duplicate(streams, 0) };
+
+        let cuda_lwe_list = CudaLweList {
+            d_vec,
+            lwe_ciphertext_count: LweCiphertextCount(1),
+            lwe_dimension,
+            ciphertext_modulus,
+        };
+        Self(cuda_lwe_list)
+    }
```

Comment:
We only had the CudaLweCiphertextList and CudaGlweCiphertextList types, and now I'm wondering if you really need to add those: can't you apply sample extraction to an array of input GLWEs and output  an array of LWEs? This is likely to be more useful for the compression of vectors of arrays, which is what we ultimately want to do I think.

Reply:
I could, but that does that make sense in this context? Compression takes a list of LWEs and outputs a single GLWE. To do what you suggest we would be running multiple compressions in parallel. Is that a use case?

Reply:
Actually this is the use case that's covered by the CPU version, if you llok at how it's done in shortint: tfhe/src/shortint/list_compression/compression.rs:19

Reply:
Oh I haven't noticed `modulus_switched_glwe_ciphertext_list` was a vector. I only saw the "Glwe" word there.

I don't see a big problem on introducing single-element versions of those ciphertext types, though. But yeah, sample extraction (and consequently, compression) would be more efficient by working with lists only. I will refactor.

Reply:
Done.

I removed `CudaGlweCiphertext` and `CudaLweCiphertext` and refactored the sample extraction rust API to take a list of GLWEs.

Reply:
@agnesLeroy actually, I'm not sure that's the use case considered in the CPU version.

Compression will indeed output a list of GLWEs. However, the unpack method used for decompression at line 120 will received a single index from the user to decompress a single ciphertext from a particular GLWE, and that's where sample extraction is used. Probably that's why there is no `glwe_sample_extraction_list.rs` in `core_crypto/algorithms`.

However, since I already did the refactor, we may now offer the capability of unpacking many ciphertexts in parallel in the GPU backend. This could be an new feature offered by the backend, but our version of the sample extraction API will need to be modified. Take a look and tell me what you think

Reply:
And now the test is broken but I will have to stop for today. Monday I fix it.